### PR TITLE
Use multi-stage build in Dockerfile

### PR DESCRIPTION
**Pull request by Gnuxie**
Fixes https://github.com/the-draupnir-project/Draupnir/issues/300

I also tried to use alpine again but rust-crypto is still fucky

with `gcompat`:
```
/node_modules/@matrix-org/matrix-sdk-crypto-nodejs/index.js:250
    throw loadError
    ^

Error: Error relocating /node_modules/@matrix-org/matrix-sdk-crypto-nodejs/matrix-sdk-crypto.linux-x64-musl.node: __register_atfork: symbol not found
    at Module._extensions..node (node:internal/modules/cjs/loader:1586:18)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
    at Module._load (node:internal/modules/cjs/loader:1104:12)
    at Module.require (node:internal/modules/cjs/loader:1311:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/node_modules/@matrix-org/matrix-sdk-crypto-nodejs/index.js:175:31)
    at Module._compile (node:internal/modules/cjs/loader:1469:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
    at Module._load (node:internal/modules/cjs/loader:1104:12) {
  code: 'ERR_DLOPEN_FAILED'
}

```

without `gcompat`:

```
/node_modules/@matrix-org/matrix-sdk-crypto-nodejs/index.js:250
    throw loadError
    ^

Error: Error loading shared library ld-linux-x86-64.so.2: No such file or directory (needed by /node_modules/@matrix-org/matrix-sdk-crypto-nodejs/matrix-sdk-crypto.linux-x64-musl.node)
    at Module._extensions..node (node:internal/modules/cjs/loader:1586:18)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
    at Module._load (node:internal/modules/cjs/loader:1104:12)
    at Module.require (node:internal/modules/cjs/loader:1311:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/node_modules/@matrix-org/matrix-sdk-crypto-nodejs/index.js:175:31)
    at Module._compile (node:internal/modules/cjs/loader:1469:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
    at Module._load (node:internal/modules/cjs/loader:1104:12) {
  code: 'ERR_DLOPEN_FAILED'
}
```

Testing was done locally by using `docker run --rm -it --network=host -v "/home/user/experiments/Draupnir/test/harness/mjolnir-data:/data" --name=draupnir draupnir-multi-stage bot --draupnir-config /data/config/production.yaml`

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -3,19 +3,24 @@
 #
 # SPDX-License-Identifier: Apache-2.0 AND AFL-3.0
 
-# We can't use alpine anymore because crypto has rust deps.
-FROM node:20-slim
+FROM alpine/git:latest as git-stamp-stage
```
Comment (Gnuxie) [inline Dockerfile:6]: is it worth doing this or should i just install git in the base image?

Comment (MTRNord) [reply Dockerfile:6]: I would argue the download overhead of the extra image is larger than an apt install which can be cached in a layer. The size of the whole image is probably larger than a single layer :thinking:

**Comment by MTRNord** [comment]
Hm I think you can then remove the version steps in the CI workflows right? :thinking:

File: Dockerfile (hunk 1 above)
Comment (Gnuxie) [reply Dockerfile:6]: oki

**Comment by Gnuxie** [comment]
> Hm I think you can then remove the version steps in the CI workflows right? 🤔

nah, we don't use the docker image in CI

**Comment by Gnuxie** [comment]
@MTRNord I think we just leave at this for now and improve later, what do you think?

**Comment by MTRNord** [comment]
> @MTRNord I think we just leave at this for now and improve later, what do you think?

sure. Whatever works for you :) making good ~~Draupnir~~Dockerimages is hard anyway.

**Pull request merged**